### PR TITLE
Add support for RulesBasedSampler

### PR DESCRIPTION
**Pull request by iNikem**
Please start reading from documentation.

/cc @theletterf 

/cc @flands I heard you had some concerned about this, so bringing this to you attention as well.

### CODE REVIEW

File: docs/advanced-config.md [hunk 1]
```diff
@@ -89,10 +89,39 @@ export OTEL_PROPAGATORS=b3multi
...
+### rules
+This sampler allows to ignore individual endpoints and drop all traces that would originate from them.
+Let's take a look at the following example configuration:
+```shell
+export OTEL_TRACES_SAMPLER=rules
+export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on
+```
 
+This configuration will result in all requests to `/healthcheck` endpoint to be excluded from monitoring.
+This includes all requests to downstream services that happen during handling this endpoint by the application.
```
Comment (theletterf) [inline docs/advanced-config.md:111]:
```suggestion
For example, the following configuration results in all requests to `/healthcheck` to be excluded from monitoring: 

```shell
export OTEL_TRACES_SAMPLER=rules
export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on
```

All requests to downstream services that happen as a consequence of calling an excluded endpoint are also excluded.
```

Comment (theletterf) [reply docs/advanced-config.md:111]: For some reason, the suggestion feature won't let me include the note below the export snippet.

File: docs/advanced-config.md [hunk 2]
```diff
@@ -89,10 +89,39 @@ export OTEL_PROPAGATORS=b3multi
...
+
+- `drop=<value>` - sampler will drop a span if its `http.url` attribute has a substring equal to the provided value.
+You can provide as many `drop` rules as you wish.
+- `default=sampler` - fallback sampler that will be used if no `drop` rule matched given span.
+Supported fallback samplers are `always_on` and `parentbased_always_on`.
+Probability samplers such as `traceidratio` are explicitly not supported.
+If several `default` rules are provided, only the last one will be in effect.
+  
+If `OTEL_TRACES_SAMPLER_ARG` is not provided or has en empty value, then no `drop` rules will be configured and `parentbased_always_on`
+sampler will be used as a default one.
```
Comment (theletterf) [inline docs/advanced-config.md:124]:
```suggestion
The value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a semicolon-separated list of rules. The following types of rules are supported:

- `drop=<value>`: The sampler drops a span if its `http.url` attribute has a substring equal to the provided value.
You can provide as many `drop` rules as you want.
- `default=sampler`: Fallback sampler used if no `drop` rule matched a  given span. Supported fallback samplers are `always_on` and `parentbased_always_on`. Probability samplers such as `traceidratio` are not supported.

> If several `default` rules are provided, only the last one will be in effect.
  
If `OTEL_TRACES_SAMPLER_ARG` is not provided or has en empty value, no `drop` rules are configured and `parentbased_always_on` sampler is as default.
```

**Review by theletterf — CHANGES_REQUESTED** [review]

**Comment by theletterf** [comment]
Previous sampler docs for `internal_root_off` were dropped out of the concern they might cause valid data to be dropped. Before we document sampler settings in full, we need to think of valid, general use cases. Happy to meet about this. @iNikem @flands

File: docs/advanced-config.md [hunk 3]
```diff
@@ -89,10 +89,35 @@ export OTEL_PROPAGATORS=b3multi
...
+Setting `otel.traces.sampler` to `internal_root_off` will drop all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`.
+This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.
+
+### rules
+This sampler allows to ignore individual endpoints and drop all traces that would originate from them.
+For example, the following configuration results in all requests to `/healthcheck` to be excluded from monitoring: 
+
+```shell
+export OTEL_TRACES_SAMPLER=rules
+export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;default=parentbased_always_on
```
Comment (jkwatson) [inline docs/advanced-config.md:108]: My question is this: is this how otel intends to support complex sampling options? If we document this, and use this format, we're going to be stuck with it, even if otel goes a different way.

Comment (iNikem) [reply docs/advanced-config.md:108]: I don't know of any plans nor opinions of OpenTelemetry for configuring "complex" samplers via environment variable. I don't know if are ready to commit this configuration. We have to give this to our customer who has requested this functionality and ask for their feedback.

File: docs/advanced-config.md [hunk 4]
```diff
@@ -89,10 +89,35 @@ export OTEL_PROPAGATORS=b3multi
...
+| `otel.traces.sampler`    | `OTEL_TRACES_SAMPLER`     | `always_on`    | Stable  | The sampler to use for tracing.	|
 
-Set `otel.traces.sampler` to `internal_root_off` to drop all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`. This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.
+Splunk Distribution of OpenTelemetry Java supports all standard samplers as provided by 
+[OpenTelemetry Java SDK Autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure#sampler).
+In addition, we add the following two samplers.
 
+### internal_root_off
+Setting `otel.traces.sampler` to `internal_root_off` will drop all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`.
+This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.
```
Comment (jkwatson) [inline docs/advanced-config.md:100]: I really really think we should *NOT* document this as something for general use.

File: docs/advanced-config.md (hunk 3 above)
Comment (jkwatson) [reply docs/advanced-config.md:108]: How would I "drop" more than one endpoint?

Comment (iNikem) [reply docs/advanced-config.md:108]: > How would I "drop" more than one endpoint?

Just add several `drop` rules

File: docs/advanced-config.md (hunk 4 above)
Comment (iNikem) [reply docs/advanced-config.md:100]: `internal_root_off` is already present in our documentation. But I will talk to Ivo and Fabrio about this.

File: docs/advanced-config.md [hunk 5]
```diff
@@ -89,10 +89,35 @@ export OTEL_PROPAGATORS=b3multi
...
+```
+All requests to downstream services that happen as a consequence of calling an excluded endpoint are also excluded.
+
+The value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a semicolon-separated list of rules. The following types of rules are supported:
+
+- `drop=<value>`: The sampler drops a span if its `http.url` attribute has a substring equal to the provided value.
+You can provide as many `drop` rules as you want.
+- `default=sampler`: Fallback sampler used if no `drop` rule matched a  given span. Supported fallback samplers are `always_on` and `parentbased_always_on`. Probability samplers such as `traceidratio` are not supported.
+
+> If several `default` rules are provided, only the last one will be in effect.
```
Comment (mateuszrzeszutek) [inline docs/advanced-config.md:118]:
```suggestion
> If several `fallback` rules are provided, only the last one will be in effect.
```

File: custom/src/main/java/com/splunk/opentelemetry/RuleBasedSamplerProvider.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+  @Override
+  public Sampler createSampler(ConfigProperties config) {
+    String configString = config.getString(SAMPLER_ARG);
+    Config samplerConfiguration = new Config(configString);
+
+    log.info("Received following rules: {}", samplerConfiguration);
+
+    RuleBasedRoutingSamplerBuilder builder =
+        RuleBasedRoutingSampler.builder(SpanKind.SERVER, samplerConfiguration.fallback);
```
Comment (mateuszrzeszutek) [inline custom/src/main/java/com/splunk/opentelemetry/RuleBasedSamplerProvider.java:45]: The docs don't mention anything about it being a `SERVER` sampler - should they?

File: docs/advanced-config.md (hunk 3 above)
Comment (jkwatson) [reply docs/advanced-config.md:108]: this should probably be documented here. It definitely wasn't clear to me from reading it.

File: docs/advanced-config.md (hunk 4 above)
Comment (theletterf) [reply docs/advanced-config.md:100]: Any news?

Comment (theletterf) [reply docs/advanced-config.md:100]: Any news?

Comment (theletterf) [reply docs/advanced-config.md:100]: Any news?

Comment (theletterf) [reply docs/advanced-config.md:100]: Any news?

Comment (theletterf) [reply docs/advanced-config.md:100]: Any news?

**Comment by iNikem** [comment]
I have removed all documentation changes. We will share them only with our beta customers who are willing to try this functionality with full disclosure that is not stable yet and may change in the future.

**Review by theletterf — APPROVED** [review]

**Review by mateuszrzeszutek — APPROVED** [review]

**Pull request merged**